### ISSUE TITLE
[EIP-0002] [FINAL] Define a Wallet Exchange Format

### ISSUE THREAD

**Issue by ivangabriele** [Electra-project/Electra-Improvement-Proposals#2]
This proposal is about defining a cross-applications format to exchange (import / export) wallets. The decisions made on EIP-0001 (#1) will impact this format.

## Why is it an issue ?

In order to ease the wallet exchange between cross-platform applications, we need to find :

- A more cross-compatible format than the [_bitkeys_ format](https://bitcointalk.org/index.php?topic=4448.0) (`wallet.dat` file).
- A short enough format to enable easy QR-Codes generation and readability.

## Proposed Solution

A space-less JSON object :

**Case A: Using HD Wallets (see #1)**

```json
{
  "chainsCount": 2,
  "masterPrivateKey": "String",
  "version": 2
}
```

In this case, the custom addresses would only be to reference non-HD addresses.

**Case B: Using non-HD (= actual) Wallets**

```json
{
  "addresses": [
    "addressOnePrivateKey",
    "addressTwoPrivateKey",
    "addressThreePrivateKey",
    "addressFourPrivateKey",
    "addressFivePrivateKey"
  ],
  "version": 2
}
```

Both of these wallets contains 5 addresses. A pure HD wallet only requires 2 values:

- the Master Node Private key,
- the chains count, that is: how many chains have been derived from this Master Node ?

while a "normal" wallet will require a list of all the private keys matching the addresses the user generated (which can be a lot !).

We need to add the case in which a "legacy" user has already randomly generated (non-HD) addresses and switch to a HD wallet. It would then look like:

**Case C: Using a HD Wallets with old non-HD addresses**

```json
{
  "chainsCount": 5,
  "customAddresses": [
    "customAddressOnePrivateKey",
    "customAddressTwoPrivateKey"
  ],
  "masterPrivateKey": "String",
  "version": 2
}
```

The `customAddresses` property is filled with the "old" addresses. Thus this wallet contains 5 HD addresses + 2 non-HD addresses.

## Proposed "Advanced" Solutions

If necessary, it would be also possible to shorten even more by :

- Creating our own "character"-separated format to getting rid of the quotes and brackets.
- Compressing this export using a compression algorithm usable in all languages and environments, using [LZM](http://rosettacode.org/wiki/LZW_compression) for example. It's left to be tested though.
- Shorten the properties names.
- Make it a pure array to avoid any property name.

**Comment by greenbigfrog** [comment]
`Creating our own "character"-separated format to getting rid of the quotes and brackets.`
Why not just use YAML?

**Comment by ivangabriele** [comment]
Because YAML is only parsable when correctly indented (thus can't be easily stringified), that's why web APIs still use JSON instead of YAML.

**Comment by ivangabriele** [comment]
Upon reflection, to have a still light but easy-to-handle solution,
I think that the best format should be a pure array JSON:

```js
[
  2, // Wallet Exchange Format version
  5, // Chains count, defaulted to 0 if it's not a HD wallet (yet)
  "masterNodePK", // defaulted to "" if it's not a HD wallet (yet)
  [
    // List of the non-HD addresses (previously randomly generated in the "legacy" wallet)
    "customAddress1PK",
    "customAddress2PK",
    ...,
  ]
]
```

**Comment by ivangabriele** [comment]
Here is the clean proposal:

---

# [EIP-0002] Wallet Exchange Format

This RFC describes the Wallet Exchange Format, that is, the unique format used to exchange an Electra wallet with any application requiring a full access to this wallet.

## Conventions used in this document

### Hierarchical Deterministic Private Keys

The Hierarchical Deterministic Wallet **MUST** follow the [Bitcoin BIP32 proposal](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki).

The Hierarchical Deterministic Master Node private key **SHOULD** have been generated via a mnemonic saved by the final user (as described in [EIP-0001](https://github.com/Electra-project/Electra-Improvement-Proposals/issues/1)) in order to allow this final user to recover their wallet without a WEF file.

### Private Keys

The private keys **MUST** be encoded in _Base58_ (Bitcoin WIF format).

The private keys described in this document **SHOULD** be ciphered by a user passphrase, following the [BIP38](https://github.com/bitcoin/bips/blob/master/bip-0038.mediawiki) procedure.

### Format

The format **MUST** be a a valid JSON following [RFC 8259](https://tools.ietf.org/html/rfc8259) specifications, including the obligation of being encoded in UTF-8 ([RFC-3629](https://tools.ietf.org/html/rfc3629)).

## Structure

The structure **MUST** be structured as above, following this exact order:

```txt
[
  VERSION_INTEGER,
  CHAINS_COUNT_INTEGER,
  HIERARCHICAL_DETERMINISTIC_MASTER_NODE_PRIVATE_KEY_STRING,
  RANDOM_ADDRESSES_PRIVATE_KEYS_STRING_ARRAY
]
```

## Content

### Version

    VERSION_INTEGER

The version integer **MUST** be equal to `2`.

Indeed, a first unofficial WEF exists in the form of a _[bitkeys](https://bitcointalk.org/index.php?topic=4448.0)_ (commonly stored as a _wallet.dat_ file). The version `1` is then de facto reserved to the legacy  _bitkeys_ format.

### Chains Count

    CHAINS_COUNT_INTEGER

The chains count integer **MUST** be equal to the number of chains derived from the Hierarchical Deterministic first derived Account from the current wallet Master Node private key.

If the wallet is Hierarchical Deterministic, this value **MUST** be greater than `0`.

If the wallet is not (yet) Hierarchical Deterministic, the chains count **MUST** be equal to `0`.

### Hierarchical Deterministic Master Node Private Key

    HIERARCHICAL_DETERMINISTIC_MASTER_NODE_PRIVATE_KEY_STRING

The Hierarchical Deterministic Master Node private key string **MUST** be equal to its WIF private key.

If the wallet is not (yet) Hierarchical Deterministic, this private key **MUST** be equal to an empty string `""`.

### Random Addresses

    RANDOM_ADDRESSES_PRIVATE_KEYS_STRING_ARRAY

The random addresses private keys array **MUST** contain all the WIF private keys as strings.

If the wallet does NOT contain any random address, the array **MUST** be specified as an empty one `[]`.

> **Note**<br>
> This field exists to ensure the backward compatibility with the legacy C++ wallet which generated random addresses.<br>
> Any new wallet SHOULD not generate any new random address anymore in order to keep the WEF data as light as possible and ease the private keys ciphering and deciphering process.

## Examples

### New HD Wallet

```json
[
  2,
  1,
  "QzVmCEXn5nVeCdfm915MFSZc6kBQk2NaDBHaGFP7n9Dshq94bp64",
  []
]
```

### Hybrid Wallet (HD + Legacy)

```json
[
  2,
  5,
  "QzVmCEXn5nVeCdfm915MFSZc6kBQk2NaDBHaGFP7n9Dshq94bp64",
  [
    "QyrDwLLxdprp8XXfcRPSpikPpFh19zXLP6NBgwVfXbLd4nTz3Sza",
    "QwsxAmZ6ndFH5CeSswZ9ybTMx5qZfJU6KM62qpQTsgMtvofQnVFz"
  ]
]
```

### Legacy Wallet

```json
[
  2,
  0,
  "",
  [
    "QyrDwLLxdprp8XXfcRPSpikPpFh19zXLP6NBgwVfXbLd4nTz3Sza",
    "QwsxAmZ6ndFH5CeSswZ9ybTMx5qZfJU6KM62qpQTsgMtvofQnVFz"
  ]
]
```

**Comment by nemo83** [comment]
I quickly scanned through this RFC, and while I agree that json format would be the best fit, I have some concerns about the implementation of this feature. 

What are we trying to solve here? Giving the possibility to someone or something else to have access to a user's wallet (regardless of how much money it contains), opens a number of issues WRT to vulnerabilities.

Can you please provide an example of what we're trying to achieve/solve? Is there any other coins doing something similar? I'd like to know more.

**Comment by ivangabriele** [comment]
@nemo83 There is no difference with the `.dat` file. It's just a change in format since `bitkeys` is a really dirty and custom format (there is not even a serious RFC describing it). As in a `.dat` file, the private keys have to be ciphered before exporting them, this is why I specified:

> The private keys described in this document SHOULD be ciphered by a user passphrase.

This is following the exact same process that Bitcoin [BIP38](https://github.com/bitcoin/bips/blob/master/bip-0038.mediawiki) recommends.

At the beginning in my proposal, I tried to explain my issues:

> **Why is it an issue ?**
> In order to ease the wallet exchange between cross-platform applications, we need to find :
> - A more cross-compatible format than the bitkeys format (wallet.dat file).
> - A short enough format to enable easy QR-Codes generation and readability.

I don't see any other easy way of exchanging the wallet between a desktop app and a mobile app, or between 2 mobile apps than via a QR code, but maybe I missed a possibility ?
